Control/Monad/Logic.hs

```diff
@@ -318,10 +323,49 @@ instance (Applicative m, F.Foldable m) => F.Foldable (LogicT m) where
 
 #endif
 
-instance T.Traversable (LogicT Identity) where
+-- A much simpler logic monad representation used to define the
+-- Traversable instance.
+newtype ML m a = ML (m (MLView m a))
+  deriving (Functor, F.Foldable, T.Traversable)
+
+data MLView m a = EmptyML | ConsML a (ML m a)
+  deriving (Functor, F.Foldable)
+
+instance T.Traversable m => T.Traversable (MLView m) where
+  traverse _ EmptyML = pure EmptyML
+  traverse f (ConsML x (ML m))
+    = liftA2 (\y ym -> ConsML y (ML ym)) (f x) (T.traverse (T.traverse f) m)
+  {- The derived instance would write the second case as
+   -
+   -   traverse f (ConsML x xs) = liftA2 ConsML (f x) (traverse @(ML m) f xs)
+   -
+   - Inlining the inner traverse gives
+   -
+   -   traverse f (ConsML x (ML m)) = liftA2 ConsML (f x) (ML <$> traverse (traverse f) m)
+   -
+   - revealing fmap under liftA2. We fuse those into a single application of liftA2,
+   - in case fmap isn't free.
+  -}
+
+toML :: Monad m => LogicT m a -> ML m a
```

Comment:
Isn't `Applicative m` enough here?

Reply:
Yeah. The thing is that it'll be awkward for people using ancient GHC and I'm not sure there's any practical benefit. It's up to you though.

Reply:
Let's use `Applicative`.

Reply:
No problem.